# Analysis code for Stankowski et al. 2023

This repository contains the code needed to conduct the analyses in our MS, currently available on BioRxiv as a preprint:

Selection on many loci drove the origin and spread of a key innovation
doi: https://doi.org/10.1101/2023.02.13.528213

Much of the code is available here, but below we provide links to other code repositories.

The code used to conduct the differential gene expression analysis is available here:
https://github.com/MartinGarlovsky/littorina_reprod_rnaseq

Calculation of population genetic summary statitics (pi, dxy, fst), was done using code in the Genomics_general repo created by Simon Martin (Many thanks for this resource!)
https://github.com/simonhmartin/genomics_general

Topology weighting was performed using the Twisst algorithm, (Martin and Van Belleghem 2017), the repo for which is also managed by Simon Martin
https://github.com/simonhmartin/twisst

We have also created a new method for analysing tpopology weights in a ternary framework. You can find the repo for the software here: 
https://github.com/HilaLifchitz/TwisstNTern/tree/main
Note this is still in development. 




